### DISCUSSION TITLE
[Bug Fix Guide] Partial Solutions for Issues with Older Kernels where KernelSU Doesn't Work and Module Functionality Fails

### DISCUSSION THREAD

**Discussion by Coconutat** [tiann/KernelSU#955] [Show and tell]
# Partial Solutions for Issues with Older Kernels where KernelSU Doesn't Work and Module Functionality Fails

## 1. Preface:

**This guide is dedicated to those who are still using older devices without custom ROMs.**

The targeted kernel versions and Android versions (I don't have devices with even older versions, so compatibility is not guaranteed for versions other than those listed below):

Kernel versions: 4.4, 4.9.
Android versions: Android 8 (O), Android 9 (P).

Prerequisite Reading: [Manually Integrate KernelSU](https://kernelsu.org/zh_CN/guide/how-to-integrate-for-non-gki.html)

Note 1: If KPROBES is not functioning properly, **you should abandon KPROBES** instead of persistently trying to make it work. **Code is static; humans are dynamic.**

***

Note 2:

For manual integration, you should disable KPROBES, as it can lead to issues such as reboots and crashes. To disable it, find the following lines in your defconfig file:

```shell
CONFIG_KPROBES=y
CONFIG_HAVE_KPROBES=y
CONFIG_KPROBE_EVENTS=y
```

Change them to:

```
# CONFIG_KPROBES is not set
# CONFIG_HAVE_KPROBES is not set
# CONFIG_KPROBE_EVENTS is not set
```

***

## 2. Fixes for Some Bugs:

**Bug 1: SU Works, but Modules Don't**

**Solution: Modify hook.c**

The code is located in:

```
security/selinux/hooks.c
```

Patch for modification: [security: selinux: allow init exec (kernelsu:ksud) under nosuid](https://github.com/sticpaper/android_kernel_xiaomi_msm8998-ksu/commit/646d0c836053ffa8743519c39fcabaecb258f00b)

***

**Bug 2: KernelSU Works on Devices Below Android 10, but SU Authorization Doesn't Work Properly, and the Module Page Shows No Overlay_FS Support**
***
**This bug has been fixed in v0.6.9.**
***
**Obsolete Content:**
> **Disclaimer**: You should not consider the following content as a final solution because it is not secure and is self-deception.

>Since KernelSU is not compatible with devices running Android 10 or earlier, it cannot handle init below Android 10 correctly. As a result, it gets intercepted by SELinux. However, at this point, you can't switch the SELinux status via adb commands, so the only option is to modify the code in selinxfs.c to forcibly set it to Permissive mode permanently. The downside of this is that it becomes highly detectable and severely compromises security.

>The code is located in:

```
security/selinux/selinuxfs.c
```

> Patch for modification: [security:selinux:selinuxfs.c: Make SELinux Mode Always in permissive mode](https://github.com/Coconutat/android_kernel_huawei_florida/commit/2bfaa7fa742119062eeff38c37705c55a3fcd72b)

***
***
**Bug 3:UMOUNT For Non GKI 4.4+ Kernels**
Thanks, by @wxt1221 https://github.com/tiann/KernelSU/pull/1060
  
cherry-pick theses commits in ksu

  + https://github.com/wxt1221/KernelSU/commit/ae8a14e708e2ee859912218d70623ee08a80b762
  + https://github.com/wxt1221/KernelSU/commit/194e09fcf11a001dab285837620ce66351fb9cf2
+ https://github.com/wxt1221/KernelSU/commit/5574642b20b0f7780fa9a2e3643eb8f99141bd68
 + https://github.com/wxt1221/KernelSU/commit/fce6f6ed6ed649df51ab468c57b9ff698fcb2701

on kernel side do the following changes in fs/namespace.c
https://github.com/wxt1221/android_kernel_oneplus_sdm845/commit/70ace81f14c97cf7896eb09c1f5a3c02a5f012ea#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9  
***
## Good Tips:
***
1. [For Kernel 3.18 follow this guide by @ananjaser1211](https://github.com/tiann/KernelSU/issues/539#issuecomment-1763398054)
2. [Compile Kernel guide For Huawei EMUI 9 /9.1.0 User](https://github.com/Coconutat/HuaweiP10-GSI-And-Modify-Or-Support-KernelSU-Tutorial/wiki/7.KernelSU%E9%80%82%E9%85%8DEMUI9%E6%88%969.1.0%E7%B3%BB%E7%BB%9F%E7%9A%84%E5%86%85%E6%A0%B8)
***
***
Other bugs are still being collected and investigated...

**Comment by anuragpindoriya** [comment]
Need help to fix this issue 

https://github.com/tiann/KernelSU/issues/959

**Reply by Coconutat** [reply]
Did you enable `CONFIG_KSU=y`?I did't see it.

**Reply by Coconutat** [reply]
And disable KPROBES!!

**Reply by anuragpindoriya** [reply]
CONFIG KPROBES is not set
CONFIG_JUMP_LABEL is not set
CONFIG_UPROBES=y
ONFIG HAVE_64BIT_ALIGNED_ACCESS is not set
CONFIG HAVE_EFFICIENT UNALIGNED_ACCESS=y
CONFIG_HAVE_KPHIES=y
CONFIG_HAVE_KRETPROBES=Y
CONFIG HAVE_ARCH_TRACEHOOK=Y
CONFIG_HAVE_DMA_CONTIGUOUS=Y
CONFIG_GENERIC_SMP_IDLE_THREAD=y

Which one

**Reply by Coconutat** [reply]
> CONFIG KPROBES is not set
> CONFIG_JUMP_LABEL is not set
> CONFIG_UPROBES=y
> ONFIG HAVE_64BIT_ALIGNED_ACCESS is not set
> CONFIG HAVE_EFFICIENT UNALIGNED_ACCESS=y
> CONFIG_HAVE_KPHIES=y
> CONFIG_HAVE_KRETPROBES=Y
> CONFIG HAVE_ARCH_TRACEHOOK=Y
> CONFIG_HAVE_DMA_CONTIGUOUS=Y
> CONFIG_GENERIC_SMP_IDLE_THREAD=y
> 
> Which one 

If you had read this post, you wouldn't have this question.

**Comment by rushiranpise** [comment]
For **Kernel 3.18** follow this guide by @ananjaser1211
https://github.com/tiann/KernelSU/issues/539#issuecomment-1763398054

**Reply by testingweb624** [reply]
Hi, I am on lineage-20 , op6T device. 

First I followed kprobe method , the su works but modules shows disabled button.

Second I followed manual one and its same. Su works but in modules section the enable/disable buttons are greyed out i.e disabled.

Any suggestion please.

Kernel version- 4.9.337
Device - oneplus 6T

**Reply by testingweb624** [reply]
Sorry please ignore this, I just missed the "bug 1" section. Following that one now.

**Comment by rushiranpise** [comment]
UMOUNT For Non GKI 4.4+ Kernels by @wxt1221 https://github.com/tiann/KernelSU/pull/1060

cherry-pick theses commits in ksu

1. https://github.com/wxt1221/KernelSU/commit/ae8a14e708e2ee859912218d70623ee08a80b762
2. https://github.com/wxt1221/KernelSU/commit/194e09fcf11a001dab285837620ce66351fb9cf2
3. https://github.com/wxt1221/KernelSU/commit/5574642b20b0f7780fa9a2e3643eb8f99141bd68
4. https://github.com/wxt1221/KernelSU/commit/fce6f6ed6ed649df51ab468c57b9ff698fcb2701


on kernel side do the following changes in fs/namespace.c
https://github.com/wxt1221/android_kernel_oneplus_sdm845/commit/70ace81f14c97cf7896eb09c1f5a3c02a5f012ea#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9

add path_umount implementation 
https://github.com/OnlyTomInSecond/android_kernel_xiaomi_sm8250/commit/e9d65c4e81d36648c512e3e11841a46a31acb301

**Reply by backslashxx** [reply]
I also did https://github.com/tiann/KernelSU/pull/1060 on 4.9, 4.14 and 4.19, works well

**Reply by backslashxx** [reply]
thanks for the cgroup headsup

**Reply by Lu5ck** [reply]
Doesn't work for 4.14, very sad. 😭😭😭

**Reply by backslashxx** [reply]
https://github.com/LineageOS/android_kernel_xiaomi_sm8250/commit/e9d65c4e81d36648c512e3e11841a46a31acb301

```
KernelSU: support umount modules for Non-GKI kernel
--- a/KernelSU/kernel/core_hook.c	2023-11-20 16:15:20.950130000 +0800
+++ b/KernelSU/kernel/core_hook.c	2023-11-20 20:40:17.052177835 +0800
@@ -489,14 +489,10 @@
 
 static void ksu_umount_mnt(struct path *path, int flags)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
 	int err = path_umount(path, flags);
 	if (err) {
 		pr_info("umount %s failed: %d\n", path->dentry->d_iname, err);
 	}
-#else
-	// TODO: umount for non GKI kernel
-#endif
 }
 
 static void try_umount(const char *mnt, bool check_mnt, int flags)

--- a/fs/namespace.c	2023-11-20 15:28:32.482832195 +0800
+++ b/fs/namespace.c	2023-11-20 20:36:43.762452629 +0800
@@ -1739,6 +1739,38 @@
 }
 #endif
 
+static int can_umount(const struct path *path, int flags)
+{
+	struct mount *mnt = real_mount(path->mnt);
+
+	if (!may_mount())
+		return -EPERM;
+	if (path->dentry != path->mnt->mnt_root)
+		return -EINVAL;
+	if (!check_mnt(mnt))
+		return -EINVAL;
+	if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */
+		return -EINVAL;
+	if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))
+		return -EPERM;
+	return 0;
+}
+
+// caller is responsible for flags being sane
+int path_umount(struct path *path, int flags)
+{
+	struct mount *mnt = real_mount(path->mnt);
+	int ret;
+
+	ret = can_umount(path, flags);
+	if (!ret)
+		ret = do_umount(mnt, flags);
+
+	/* we mustn't call path_put() as that would clear mnt_expiry_mark */
+	dput(path->dentry);
+	mntput_no_expire(mnt);
+	return ret;
+}
 /*
  * Now umount can handle mount points as well as block devices.
  * This is important for filesystems which use unnamed block devices.

```

tested working on 4.9, 4.14 and 4.19

**Reply by Ardjlon** [reply]
Is this supposed to work so that the modules are not detected?
Because I am doing test for an app, but it always detect modules.img (4.14)

**Reply by Ardjlon** [reply]
The dmsg messages show that it dismounts correctly, but the app still detects the modules

[a comment by an automated account is not shown]
